lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+// Copyright (c) 2012-2017, Najjar Chedy
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+var _ = require('underscore-node');
+
+function ControlFlowGraph(rules){
+    this.rules = rules;
+    
+}
+
+function seperateCodeFromData(asmArr) {
+    
+    var isCode = function(asmArr, index){
+        return (((asmArr[index].source !== null) &&
+                (asmArr[index+1].source !== null)) ||
+                ((asmArr[index].source === null) &&
+                 (asmArr[index+1].source !== null)) ||
+                 ((asmArr[index].source !== null) &&
```

Comment:
If I read this right, this function returns `true` iff:

* `index` and `index+1` are non-null
* `index` is null but `index+1` is non-null
* `index` is non-null but `index+` is null

The only case not explicitly checked for is both `index` and `index+` being null, which returns `false`. So I think the whole thing can be replaced with something like:
```
var isCode = function (asmArr, index) {
  return asmArr[index].source !== null || asmArr[index+1].source !== null;
};
```

Unless I'm missing something?  Seem it doesn't handle `index` being the last element too? With a simple check it could be made to I think.

Reply:
the stupidest mistake  I made.